File: src/api/Falchion.Villains.Vault.Api/Models/RaceStats.cs
using System.ComponentModel.DataAnnotations;
using System.Text.Json;

namespace Falchion.Villains.Vault.Api.Models
{
    public class AgeGroupItem
    {
        public string AgeGroupLabel { get; set; } = string.Empty;
        public int Count { get; set; }
        public List<int> AgeRange { get; set; } = new List<int> { 0, 0 };
        public TimeSpan AverageNetTime { get; set; } = TimeSpan.Zero;
        public TimeSpan AveragePace { get; set; } = TimeSpan.Zero;
        public TimeSpan MedianPace { get; set; } = TimeSpan.Zero;
        public int DNFCount { get; set; }
    }
    public class SplitTimeStats
    {
        /// <summary>
        /// The label for this split as it appears in the race results.
        /// Examples: "5K", "10K", "Half Marathon", "5 Mile", "11.5 Mile"
        /// </summary>
        public string Label { get; set; } = string.Empty;

        /// <summary>
        /// The average pace achieved from the previous split (start if this is the first split) to the split marker.
        /// </summary>
        public TimeSpan AveragePace { get; set; } = TimeSpan.Zero;

        /// <summary>
        /// The median pace achieved from the previous split (start if this is the first split) to the split marker.
        /// </summary>
        public TimeSpan MedianPace { get; set; } = TimeSpan.Zero;

        /// <summary>
        /// The total distance from the start line to the split marker. If the finish split then this will be the total race distance.
        /// </summary>
        public double TotalDistanceToSplitInMiles { get; set; }

        /// <summary>
        /// The total distance from the previous split (start if this is the first split) to the split marker. If the finish split then this will be from the last available split to the finish line.
        /// </summary>
        public double SegmentDistanceInMiles { get; set; }

        /// <summary>
        /// The total number of runners that have no data for the split. If this is the finish split then this will be all of the DNF runners.
        /// </summary>
        public int Misses { get; set; } = 0;
    }
    public class RaceStats
    {
        /// <summary>
        /// The total number of all runners that participated including DNF, Hand Cycle, Push Rim, Duo, and Unknown
        /// </summary>
        public int TotalRunners { get; set; }

        /// <summary>
        /// The total number of all male runners that participated including DNF, Hand Cycle, Push Rim. Does not include Duo runners since we cannot determine their gender.
        /// </summary>
        public int MaleRunners { get; set; }

        /// <summary>
        /// The total number of all female runners that participated including DNF, Hand Cycle, Push Rim. Does not include Duo runners since we cannot determine their gender.
        /// </summary>
        public int FemaleRunners { get; set; }

        /// <summary>
        /// The total number of runners with unknown gender. Includes DNF runners whose gender could not be determined.
        /// </summary>
        public int UnknownRunners { get; set; }

        /// <summary>
        /// The total number of all runners, including DNF. Hand Cycle, Push Rim, Duo, and Unknown runner types are not included.
        /// </summary>
        public int RunnerTypeRunner { get; set; }

        /// <summary>
        /// The total number of all Push Rim runners that participated. Does not include DNF as we can't determine their runner type so we assume runner.
        /// </summary>
        public int RunnerTypePushRim { get; set; }

        /// <summary>
        /// The total number of all Hand Cycle runners that participated. Does not include DNF as we can't determine their runner type so we assume runner.
        /// </summary>
        public int RunnerTypeHandCycle { get; set; }

        /// <summary>
        /// The total number of all Duo runners that participated. Does not include DNF as we can't determine their runner type so we assume runner.
        /// </summary>
        public int RunnerTypeDuo { get; set; }

        /// <summary>
        /// The total number of DNF runners that participated. DNF runners are stored in a special division and excluded from normal queries, so we have to query them separately.
        /// </summary>
        public int DNFCount { get; set; }

        /// <summary>
        /// For each available split time, break it up into segments so if you have 4 splits there will be five segments where Split1 is from 0mi to Split1.Distance and the last split is from SplitN.Distance to the finish which would be the total race distance.
        /// </summary>
        public List<SplitTimeStats> Splits { get; set; } = new List<SplitTimeStats>();

        /// <summary>
        /// The total number of runners with an overall pace slower than 16 min/mile.
        /// </summary>
        public int RunnersOver16minPace { get; set; }

        /// <summary>
        /// How long was the start line open (from the first to the last starter crossing)?
        /// </summary>
        public TimeSpan LaunchTime { get; set; }

        /// <summary>
        /// Calculate the launch congestion factor as the ratio of total runners to launch time in minutes.
        /// </summary>
        public decimal LaunchCongestionFactor { get; set; }

        /// <summary>
        /// How long was the finish line open (from the first to the last finisher crossing)?
        /// Don't include DNF runners in this calculation since they never crossed the finish line.
        /// Don't include Hand Cycle and Push Rim runners since they have different pace profiles and will potentially finish considerably earlier than the first runners.
        /// </summary>
        public TimeSpan LandingTime { get; set; }

        /// <summary>
        /// Calculate the landing congestion factor as the ratio of total runners to landing time in minutes.
        /// </summary>
        public decimal LandingCongestionFactor { get; set; }

        /// <summary>
        /// The slowest finisher (excluding DNF) with a valid net time.
        /// </summary>
        public long? SlowestFinisherResultId { get; set; }

        /// <summary>
        /// The last person to cross the finish line (excluding DNF).
        /// </summary>
        public long? LastFinisherResultId { get; set; }

        /// <summary>
        /// Stats for each male age group. Age groups are broken down to 5 year increments (e.g., 0-4, 5-9, 10-14, etc.) with the final group being 80+. Each age group includes the count of runners in that group, the average net time, and the average pace.
        /// </summary>
        public List<AgeGroupItem> MaleAgeGroupStats { get; set; } = new List<AgeGroupItem>();

        /// <summary>
        /// Stats for each female age group. Age groups are broken down to 5 year increments (e.g., 0-4, 5-9, 10-14, etc.) with the final group being 80+. Each age group includes the count of runners in that group, the average net time, and the average pace.
        /// </summary>
        public List<AgeGroupItem> FemaleAgeGroupStats { get; set; } = new List<AgeGroupItem>();

        /// <summary>
        /// Serializes this instance to JSON string.
        /// </summary>
        public string ToJson()
        {
            return JsonSerializer.Serialize(this, new JsonSerializerOptions { PropertyNamingPolicy = JsonNamingPolicy.CamelCase });
        }

        /// <summary>
        /// Deserializes a JSON string to a RaceStats instance.
        /// </summary>
        public static RaceStats? FromJson(string? json)
        {
            if (string.IsNullOrWhiteSpace(json))
                return null;

            return JsonSerializer.Deserialize<RaceStats>(json, new JsonSerializerOptions { PropertyNamingPolicy = JsonNamingPolicy.CamelCase });
        }
    }
}

File: src/app/models/races/RaceStats.ts
/**
 * Age group statistics for a specific age range
 */
export interface AgeGroupItem {
	/** Label for the age group (e.g., "20-24", "80+") */
	ageGroupLabel: string;
	
	/** Number of runners in this age group */
	count: number;
	
	/** Age range for this group */
	ageRange: {
		min: number;
		max: number;
	};
	
	/** Average net time for this age group (ISO 8601 duration format) */
	averageNetTime: string;
	
	/** Average pace for this age group (ISO 8601 duration format) */
	averagePace: string;
	
	/** Median pace for this age group (ISO 8601 duration format) */
	medianPace: string;
	
	/** Number of DNF runners in this age group */
	dnfCount: number;
}

/**
 * Extended split time information with statistics
 */
export interface SplitTimeStats {
    /** The label for this split as it appears in the race results. Examples: "5K", "10K", "Half Marathon", "5 Mile", "11.5 Mile" */
    label: string;

	/** Average pace from previous split to this split marker (ISO 8601 duration format) */
	averagePace: string;
	
	/** Median pace from previous split to this split marker (ISO 8601 duration format) */
	medianPace: string;
	
	/** The total distance from the start line to the split marker. If the finish split then this will be the total race distance. */
	totalDistanceToSplitInMiles: number;

    /** The total distance from the previous split (start if this is the first split) to the split marker. If the finish split then this will be from the last available split to the finish line. */
    segmentDistanceInMiles: number;
	
	/** Number of runners with no data for this split (DNF if finish split) */
	misses: number;
}

/**
 * Comprehensive race statistics
 */
export interface RaceStats {
	/** Total number of all participants including DNF, Hand Cycle, Push Rim, Duo, and Unknown */
	totalRunners: number;
	
	/** Total number of male runners (excluding Duo since gender cannot be determined) */
	maleRunners: number;
	
	/** Total number of female runners (excluding Duo since gender cannot be determined) */
	femaleRunners: number;

	/** Total number of runners with unknown gender */
	unknownRunners: number;
	
	/** Total number of standard runners (excluding Hand Cycle, Push Rim, Duo, and Unknown) */
	runnerTypeRunner: number;
	
	/** Total number of Push Rim runners */
	runnerTypePushRim: number;
	
	/** Total number of Hand Cycle runners */
	runnerTypeHandCycle: number;
	
	/** Total number of Duo runners */
	runnerTypeDuo: number;
	
	/** Total number of DNF (Did Not Finish) runners */
	dnfCount: number;
	
	/** Split time statistics for each segment of the race */
	splits?: SplitTimeStats[];
	
	/** Number of runners with overall pace slower than 16 min/mile */
	runnersOver16minPace: number;
	
	/** Duration from first to last starter crossing the start line (ISO 8601 duration format) */
	launchTime: string;
	
	/** Launch congestion factor (ratio of total runners to launch time in minutes) */
	launchCongestionFactor: number;
	
	/** Duration from first to last finisher crossing the finish line (ISO 8601 duration format) */
	landingTime: string;
	
	/** Landing congestion factor (ratio of total runners to landing time in minutes) */
	landingCongestionFactor: number;
	
	/** Result ID of the slowest finisher (excluding DNF) */
	slowestFinisherResultId?: number;
	
	/** Result ID of the last person to cross the finish line (excluding DNF) */
	lastFinisherResultId?: number;
	
	/** Age group statistics for male runners */
	maleAgeGroupStats?: AgeGroupItem[];
	
	/** Age group statistics for female runners */
	femaleAgeGroupStats?: AgeGroupItem[];
}
